libfwupdplugin/fu-plugin.h

```diff
@@ -61,6 +72,291 @@ typedef enum {
+				   GBytes *blob_fw,
+				   FuProgress *progress,
+				   FwupdInstallFlags flags,
+				   GError **error);
+	/**
+	 * unlock:
+	 * @self: a #FuPlugin
+	 * @dev: a device
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Unlocks the device for writes.
+	 *
+	 * Since: 0.8.0
+	 **/
+	gboolean (*unlock)(FuPlugin *self, FuDevice *device, GError **error);
+	/**
+	 * activate:
+	 * @self: a #FuPlugin
+	 * @dev: a device
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Activates the new firmware on the device.
+	 *
+	 * This is intended for devices that it is not safe to immediately activate
+	 * the firmware.  It may be called at a more convenient time instead.
+	 *
+	 * Since: 1.2.6
+	 **/
+	gboolean (*activate)(FuPlugin *self,
+			     FuDevice *device,
+			     FuProgress *progress,
+			     GError **error);
+	/**
+	 * attach:
+	 * @self: a #FuPlugin
+	 * @dev: a device
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Swaps the device from bootloader mode to runtime mode.
+	 *
+	 * Since: 1.0.2
+	 **/
+	gboolean (*attach)(FuPlugin *self, FuDevice *device, FuProgress *progress, GError **error);
+	/**
+	 * detach:
+	 * @self: a #FuPlugin
+	 * @dev: a device
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Swaps the device from runtime mode to bootloader mode.
+	 *
+	 * Since: 1.7.0
+	 **/
+	gboolean (*detach)(FuPlugin *self, FuDevice *device, FuProgress *progress, GError **error);
+	/**
+	 * prepare:
+	 * @self: a #FuPlugin
+	 * @dev: a device
+	 * @flags: install flags
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Prepares the device to receive an update.
+	 *
+	 * Since: 1.7.0
+	 **/
+	gboolean (*prepare)(FuPlugin *self,
+			    FuDevice *device,
+			    FwupdInstallFlags flags,
+			    GError **error);
+	/**
+	 * cleanup
+	 * @self: a #FuPlugin
+	 * @dev: a device
+	 * @flags: install flags
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Cleans up the device after receiving an update.
+	 *
+	 * Since: 1.7.0
+	 **/
+	gboolean (*cleanup)(FuPlugin *self,
+			    FuDevice *device,
+			    FwupdInstallFlags flags,
+			    GError **error);
+	/**
+	 * composite_prepare
+	 * @self: a #FuPlugin
+	 * @devices: (element-type FuDevice): array of devices
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Function run before updating group of composite devices.
+	 *
+	 * Since: 1.0.9
+	 **/
+	gboolean (*composite_prepare)(FuPlugin *self, GPtrArray *devices, GError **error);
+	/**
+	 * composite_cleanup
+	 * @self: a #FuPlugin
+	 * @devices: (element-type FuDevice): array of devices
+	 * @error: (nullable): optional return location for an error
+	 *
+	 * Function run after updating group of composite devices.
+	 *
+	 * Since: 1.0.9
+	 **/
+	gboolean (*composite_cleanup)(FuPlugin *self, GPtrArray *devices, GError **error);
+	/*< private >*/
+	gpointer padding[9];
+} FuPluginVfuncs;
```

Comment:
hey I got an idea here.  How about you include a member here for `build_hash`?  Then you can stop having to have every plugin call set_build_hash and it can be part of the vtable.

Reply:
ack, good idea. will try.

Reply:
That works really well. Done.